src/main.rs

```diff
@@ -187,10 +237,97 @@ fn byte_offset_to_point(file_contents: &Rope, byte_offset: usize) -> Point {
+        end: byte_offset_to_position(file_contents, range.end),
+    }
+}
+
+fn tree_sitter_range_to_lsp_range(file_contents: &Rope, range: tree_sitter::Range) -> Range {
+    byte_offset_range_to_lsp_range(file_contents, range.start_byte..range.end_byte)
+}
+
+fn get_text_document_edits(
+    old_contents: &Rope,
+    new_contents: &Rope,
+    uri: &Url,
+) -> TextDocumentEdit {
+    let old_contents_as_str: Cow<'_, str> = old_contents.clone().into();
+    let new_contents_as_str: Cow<'_, str> = new_contents.clone().into();
+    let (_, diffs) = text_diff::diff(&old_contents_as_str, &new_contents_as_str, "");
+    let mut old_bytes_seen = 0;
+    let mut just_seen_remove: Option<String> = None;
+    let mut edits: Vec<TextEdit> = Default::default();
+    for diff in diffs {
+        if just_seen_remove.is_some() && !matches!(&diff, Difference::Add(_)) {
+            edits.push(TextEdit {
+                range: byte_offset_range_to_lsp_range(
+                    old_contents,
+                    old_bytes_seen - just_seen_remove.as_ref().unwrap().len()..old_bytes_seen,
+                ),
+                new_text: Default::default(),
+            });
+            just_seen_remove = None;
+        }
+        match diff {
+            Difference::Same(diff) => old_bytes_seen += diff.len(),
+            Difference::Rem(diff) => {
+                old_bytes_seen += diff.len();
+                just_seen_remove = Some(diff);
+            }
+            Difference::Add(diff) => {
+                if just_seen_remove.is_some() {
+                    edits.push(TextEdit {
+                        range: byte_offset_range_to_lsp_range(
+                            old_contents,
+                            old_bytes_seen - just_seen_remove.as_ref().unwrap().len()
+                                ..old_bytes_seen,
+                        ),
+                        new_text: diff,
+                    });
+                    just_seen_remove = None;
+                } else {
+                    edits.push(TextEdit {
+                        range: byte_offset_range_to_lsp_range(
+                            old_contents,
+                            old_bytes_seen..old_bytes_seen,
+                        ),
+                        new_text: diff,
+                    });
+                }
+            }
+        }
+    }
+    if let Some(just_seen_remove) = just_seen_remove.as_ref() {
+        edits.push(TextEdit {
+            range: byte_offset_range_to_lsp_range(
+                old_contents,
+                old_bytes_seen..old_bytes_seen + just_seen_remove.len(),
+            ),
+            new_text: Default::default(),
+        });
+    }
+    TextDocumentEdit {
+        text_document: OptionalVersionedTextDocumentIdentifier {
+            uri: uri.clone(),
+            version: None,
+        },
+        edits: edits.into_iter().map(OneOf::Left).collect(),
+    }
+}
+
+fn get_uri_from_arguments(arguments: &[serde_json::Value]) -> Url {
```

Comment:
It seems like it's up to the LSP client to know what "arguments" type to send for this "command"?

This is mimicking what I saw for the ESLint LSP server I have set up which is that it was passing something like `arguments: [{ uri: "...", version: 5 }]` (we should probably be keeping track of these file "versions" but haven't really wrapped my head around that yet)